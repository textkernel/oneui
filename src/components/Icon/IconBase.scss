@import '../../themes/oneui/oneui';

.IconBase {
    display: inline-flex;
    align-self: center;
    fill: currentColor;

    &--isPrimary {
        fill: var(--color-primary);
    }

    &__svg {
        position: relative;
        top: 0.125em;
        height: 1em;
        transition: fill var(--transition-duration);
    }

    &--margin {
        @each $side in (top, right, bottom, left) {
            &_#{ $side } {
                margin-#{ $side }: 0.25em;
            }
        }
    }
}
