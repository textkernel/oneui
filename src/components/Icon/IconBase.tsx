import React, { forwardRef } from 'react';
import { bem } from '../../utils';
import styles from './IconBase.scss';
import { Context } from '../../constants';

export interface IconProps extends React.HTMLAttributes<HTMLDivElement> {
    /** The icon context (e.g. brand, primary, bad, good etc. - defaults to brand) */
    context?: Context;
    /** Adds margin between a given side of the icon and other content */
    margin?: 'top' | 'right' | 'bottom' | 'left';
    /** Absolute size for this icon (size in pixels, aspect ratio is 1:1).
     If not defined, icon will scale and align itself with text. */
    size?: number;
    /** If true, width will set to be automatic */
    preserveAspectRatio?: boolean;
    /** Optional icon title */
    title?: string;
}

export interface Props extends IconProps {
    /** The SVG content */
    children: NotEmptyReactNode;
    /** The SVG viewbox */
    viewBox: string;
}

const { block, elem } = bem('IconBase', styles);

const adjustSize = (preserveAspectRatio: boolean, size?: number) => {
    if (!size) {
        if (!preserveAspectRatio) {
            return { width: '1em' };
        }
        return { width: 'auto' };
    }

    const adjustedSize = Math.max(0, size);

    return {
        top: 'auto',
        width: preserveAspectRatio ? 'auto' : adjustedSize,
        height: adjustedSize,
    };
};

export const IconBase = forwardRef<HTMLDivElement, Props>(
    (
        { children, context, margin, size, preserveAspectRatio = false, title, viewBox, ...rest },
        ref
    ) => {
        return (
            <div ref={ref} {...rest} {...block({ margin, context, ...rest })}>
                <svg
                    {...elem('svg')}
                    aria-labelledby={title ? 'title' : null}
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox={viewBox}
                    style={adjustSize(!!preserveAspectRatio, size)}
                    role="img"
                >
                    {!!title && <title>{title}</title>}
                    {children}
                </svg>
            </div>
        );
    }
);

IconBase.displayName = 'IconBase';
