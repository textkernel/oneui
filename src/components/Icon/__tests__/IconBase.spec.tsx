import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { IconBase } from '../IconBase';

describe('<IconBase> that renders an SVG wrapper with all options included', () => {
    it('should render nothing if no children are provided', () => {
        const { container } = render(<IconBase viewBox="0 0 100 100">{}</IconBase>);

        expect(container).toBeEmptyDOMElement();
    });
    it('should render children when it is 0', () => {
        const { container } = render(<IconBase viewBox="0 0 100 100">{0}</IconBase>);

        expect(container).not.toBeEmptyDOMElement();
    });
    it('should render a default icon', () => {
        const view = render(
            <IconBase viewBox="0 0 100 100" title="Icon base" size={15} margin="right" isPrimary>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(view.container).toMatchSnapshot();

        expect(screen.getByTitle('Icon base')).toBeInTheDocument();
    });
    it('should correct negative sizes', () => {
        const view = render(
            <IconBase viewBox="0 0 100 100" size={-1}>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(view.container).toMatchSnapshot();
        const img = screen.getByRole('img');

        expect(img).toBeInTheDocument();
        expect(img.style.height).toBe('0px');
        expect(img.style.width).toBe('0px');
        expect(screen.queryByTitle('Icon base')).not.toBeInTheDocument();
    });
    it('should not apply proportional styles if no size provided', () => {
        const view = render(
            <IconBase viewBox="0 0 100 100">
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(view.container).toMatchSnapshot();
        const img = screen.getByRole('img');

        expect(img).toBeInTheDocument();
        expect(img.style.height).toBe('');
        expect(img.style.width).toBe('1em');
    });
    it('should apply proportional styles if size is provided', () => {
        const mockSize = 48;
        const view = render(
            <IconBase viewBox="0 0 100 100" size={mockSize}>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(view.container).toMatchSnapshot();
        const img = screen.getByRole('img');

        expect(img).toBeInTheDocument();
        expect(img.style.height).toBe('48px');
        expect(img.style.width).toBe('48px');
    });
    it('should apply height only to styles if size is provided with preserveAspectRatio', () => {
        const mockSize = 48;
        const view = render(
            <IconBase viewBox="0 0 100 100" size={mockSize} preserveAspectRatio>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(view.container).toMatchSnapshot();
        const img = screen.getByRole('img');

        expect(img).toBeInTheDocument();
        expect(img.style.height).toBe('48px');
        expect(img.style.width).toBe('auto');
    });
});
