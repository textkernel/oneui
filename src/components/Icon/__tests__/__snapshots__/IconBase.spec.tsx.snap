// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconBase> that renders an SVG wrapper with all options included should apply height only to styles if size is provided with preserveAspectRatio 1`] = `
<div>
  <div
    class="IconBase"
  >
    <svg
      class="IconBase__svg"
      role="img"
      style="width: auto; height: 48px;"
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="50"
        cy="50"
        r="50"
      />
    </svg>
  </div>
</div>
`;

exports[`<IconBase> that renders an SVG wrapper with all options included should apply proportional styles if size is provided 1`] = `
<div>
  <div
    class="IconBase"
  >
    <svg
      class="IconBase__svg"
      role="img"
      style="width: 48px; height: 48px;"
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="50"
        cy="50"
        r="50"
      />
    </svg>
  </div>
</div>
`;

exports[`<IconBase> that renders an SVG wrapper with all options included should correct negative sizes 1`] = `
<div>
  <div
    class="IconBase"
  >
    <svg
      class="IconBase__svg"
      role="img"
      style="width: 0px; height: 0px;"
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="50"
        cy="50"
        r="50"
      />
    </svg>
  </div>
</div>
`;

exports[`<IconBase> that renders an SVG wrapper with all options included should not apply proportional styles if no size provided 1`] = `
<div>
  <div
    class="IconBase"
  >
    <svg
      class="IconBase__svg"
      role="img"
      style="width: 1em;"
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="50"
        cy="50"
        r="50"
      />
    </svg>
  </div>
</div>
`;

exports[`<IconBase> that renders an SVG wrapper with all options included should render a default icon 1`] = `
<div>
  <div
    class="IconBase IconBase--margin_right IconBase--isPrimary"
  >
    <svg
      aria-labelledby="title"
      class="IconBase__svg"
      role="img"
      style="width: 15px; height: 15px;"
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Icon base
      </title>
      <circle
        cx="50"
        cy="50"
        r="50"
      />
    </svg>
  </div>
</div>
`;
